core/include/prometheus/gauge_builder.h

```diff
@@ -14,20 +14,61 @@ namespace detail {
 class GaugeBuilder;
 }
 
+/// \brief Return a builder to allow configuration of a family of gauges.
+///
+/// A family groups a set of gauges under a common name tagged with labels.
+///
+/// Example usage:
+///
+/// \code
+/// auto registry = std::make_shared<Registry>();
+/// auto& counter_family = prometheus::BuildGauge()
+///                            .Name("some_name")
+///                            .Help("Additional description.")
+///                            .Labels({{"key", "value"}})
+///                            .Register(*registry);
+///
+/// ...
+/// \endcode
+///
+/// Every time series is uniquely identified by its metric name and a set of
+/// key-value pairs, also known as labels. Use GaugeBuilder::Name(const
+/// std::string&) to set the metric name. Use GaugeBuilder::Label(const
+/// std::map<std::string, std::string>&) to set a label. Use
+/// GaugeBuilder::Help(const std::string&) to set an additional description.
+///
+/// It is required to follow the syntax of metric names and labels given by:
+/// https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+///
+/// The following metric and label conventions are not required for using
+/// Prometheus, but can serve as both a style-guide and a collection of best
+/// practices: https://prometheus.io/docs/practices/naming/
+///
+/// Finally the gauge family must be registered as a collector with
+/// GaugeBuilder::Register(Registry&). The gauge family fullfills the
+/// Collectable interface to expose the collected metrics. The registration is a
+/// key part needed to scrape metrics.
 detail::GaugeBuilder BuildGauge();
```

Comment:
During documenting the API I wondered why this function exists at all? Essentially the coding style

```
prometheus::BuildGauge()
                           .Name("some_name")
```
can also be realized with

```
detail::GaugeBuilder{}
   .Name(...)
   ...
```

It also brings me to the next point: Why is `GaugeBuilder` in namespace detail and thus an implementation detail? From the current usage it is an integral part of the API a user must use. So it should be either lifted into the prometheus namespace.

Or - if it is really desired to make `GaugeBuilder` an implementation detail - change `BuildGauge()` to something like `Family<Gauge>& RegisterGaugeFamily(labels, name, help, registry);` With the latter API the `GaugeBuilder` is truly an implementation detail because it is hidden from the user.

Reply:
But what is the benefit of this indirection vs. creating a Family and register the Family object?

But regardless of this, if the builders are an implementation detail I would move them to the include/detail directory:

https://github.com/jupp0r/prometheus-cpp/pull/214